Keas.Mvc/Controllers/TeamAdminController.cs

```diff
@@ -402,147 +431,333 @@ public async Task<IActionResult> UploadCSV(IFormFile file)
+                        else
+                        {
+                            if (somethingSaved)
+                            {
+                                errorCount += 1;
+                                transaction.Rollback();
+                                if (!string.IsNullOrWhiteSpace(r.KerbUser))
+                                {
+                                    var local = _context.Set<User>()
+                                        .Local
+                                        .FirstOrDefault(entry => entry.Id.Equals(r.KerbUser));
+
+                                    // check if local is not null 
+                                    if (local != null) // I'm using a extension method
+                                    {
+                                        // detach
+                                        _context.Entry(local).State = EntityState.Detached;
                                     }
-                                    if (import)
+
+                                    var localPerson = _context.Set<Person>().Local
+                                        .FirstOrDefault(entry => entry.UserId.Equals(r.KerbUser));
+                                    if (localPerson != null) // I'm using a extension method
                                     {
-                                        assignment.PersonId = person.Id;
-                                        assignment.KeySerialId = serial.Id;
-                                        assignment.RequestedById = User.Identity.Name;
-                                        assignment.RequestedByName = User.GetNameClaim();                                    
+                                        // detach
+                                        _context.Entry(localPerson).State = EntityState.Detached;
                                     }
+
                                 }
                             }
                         }
-                        await _context.SaveChangesAsync();
+
+                        if (!result.Success)
+                        {
+                            result.Messages = new List<string>();
+                        }
+                        resultsView.Add(result);
                     }
                 }
             }
-            Message = string.Format("Successfully loaded {0} new keys, {1} new keySerials, {2} new and {3} reactivated team members, and {4} new assignments recorded. {5}", keyCount, serialCount, peopleCount, reactivatedCount, assignmentCount, warning.ToString());
-            return RedirectToAction("Index");
+
+            Message = $"Successfully loaded {keyCount} new keys, {serialCount} new keySerials, {peopleCount} new or reactivated team members, and {assignmentCount} new assignments recorded.";
+            if (errorCount > 0)
+            {
+                ErrorMessage = $"{errorCount} rows not imported due to errors.";
+            }
+            return View(resultsView);
```

Comment:
@jSylvestre maybe not a deal breaker but this method is way way too long with too many nested if/else statements.  It could really do to break some of the sections into small private methods